edb/tools/gen_rust_ast.py

```diff
@@ -0,0 +1,274 @@
+
+        union_types.append(
+            ASTUnion(name=name, variants=variants, for_composition=True)
+        )
+
+        if cls.typ.__abstract_node__:
+            field_type = name
+        else:
+            field_type = f'Option<{name}>'
+
+        fields += f'    #[py_child]\n' f'    pub {kind_name}: {field_type},\n'
+
+    return (
+        f'\n{doc_comment}'
+        + f'#[derive(Debug, Clone)]\n'
+        + f'#[cfg_attr(feature = "python", derive(IntoPython))]\n'
+        + f'pub struct {cls.name} {"{"}\n'
+        + fields
+        + '}\n'
+    ).replace('{\n}', r'{}')
+
+
+def codegen_enum(name: str, cls: typing.Type) -> str:
+    fields = ''
+    for member in cls._member_names_:
+        fields += f'    {member},\n'
+
+    if cls.__module__ == 'edb.edgeql.ast':
+        cls_path = f'qlast.{cls.__name__}'
+    elif cls.__module__ == 'edb.edgeql.qltypes':
+        cls_path = f'qltypes.{cls.__name__}'
+    else:
+        raise LookupError(
+            'we only support generating AST from qlast and qltypes modules'
+        )
+
+    return (
+        '\n#[derive(Debug, Clone)]\n'
+        + f'#[cfg_attr(feature = "python", derive(IntoPython))]\n'
+        + f'#[py_enum({cls_path})]\n'
+        + f'pub enum {name} {"{"}\n'
+        + fields
+        + '}\n'
+    )
+
+
+def quote_rust_ident(name: str) -> str:
+    if name in {'type', 'where', 'ref', 'final', 'abstract'}:
+        return 'r#' + name
+    return name
+
+
+def title_case(name: str) -> str:
+    return name[0].upper() + name[1:]
+
+
+def codegen_union(union: ASTUnion) -> str:
+    fields = ''
+    for arg in union.variants:
+        if isinstance(arg, str):
+            fields += f'    {arg},\n'
+        else:
+            typ = translate_type(arg, '???', union.for_composition)
+            fields += f'    {arg.__name__}({typ}),\n'
+
+    attr = 'py_child' if union.for_composition else 'py_union'
+
+    return (
+        '\n#[derive(Debug, Clone)]\n'
+        f'#[cfg_attr(feature = "python", derive(IntoPython))]\n'
+        f'#[{attr}]\n'
+        f'pub enum {union.name} {"{"}\n{fields}{"}"}\n'
+    )
+
+
+def translate_type(
+    typ: typing.Type, union_name: str, for_composition: bool
+) -> str:
+    params = [
+        translate_type(param, union_name, for_composition)
+        for param in typing_inspect.get_args(typ)
+    ]
+
+    if typing_inspect.is_union_type(typ):
+
+        if hasattr(typ, '_name') and typ._name == 'Optional':
+            return f'Option<{params[0]}>'
+
+        union_types.append(
+            ASTUnion(
+                name=union_name,
+                variants=typing_inspect.get_args(typ),
+                for_composition=for_composition,
+            )
+        )
+        return union_name
+
+    if typing_inspect.is_generic_type(typ) and hasattr(typ, '_name'):
+
+        if typ._name in ('List', 'Sequence'):
+            return f'Vec<{params[0]}>'
+
+        if typ._name == 'Dict':
+            return f'HashMap<{params[0]}, {params[1]}>'
```

Comment:
We need to make sure to use an order preserving map. Probably https://docs.rs/indexmap/latest/indexmap/.

Reply:
Let's stick with [`std::collections::BTreeMap`](https://doc.rust-lang.org/std/collections/struct.BTreeMap.html) for now. It has a bit worse performance, but it's in `std` so we avoid a dependency.

It's easy to switch anyway.

Reply:
BTreeMap is not order-preserving. Its sorted.

(switch might not be easy, as there are different traits on the key of BTreeMap vs Hash/Indexmap, although they are probably all strings in AST)

Reply:
Oh, that easy to mix up :D It's indexmap then.